semantic/poly_types.go

```diff
@@ -0,0 +1,778 @@
+	}
+}
+func (o object) resolveType(kinds map[Tvar]Kind) (Type, error) {
+	return o.krecord.resolveType(kinds)
+}
+func (o object) MonoType() (Type, bool) {
+	return o.krecord.MonoType()
+}
+func (o object) resolvePolyType(kinds map[Tvar]Kind) (PolyType, error) {
+	return o.krecord.resolvePolyType(kinds)
+}
+func (o object) Equal(t PolyType) bool {
+	switch t := t.(type) {
+	case object:
+		if len(o.krecord.properties) != len(t.krecord.properties) ||
+			!o.krecord.lower.equal(t.krecord.lower) ||
+			!o.krecord.upper.equal(t.krecord.upper) {
+			return false
+		}
+		for k, p := range o.krecord.properties {
+			op, ok := t.krecord.properties[k]
+			if !ok || !p.Equal(op) {
+				return false
+			}
+		}
+		return true
+	default:
+		return false
+	}
+}
+
+func (o object) KindConstraint() Kind {
+	return o.krecord
+}
+
+type KindConstrainter interface {
+	KindConstraint() Kind
+}
+
+type KClass struct{}
+
+func (k KClass) freeVars(c *Constraints) TvarSet { return nil }
+func (k KClass) substituteKind(tv Tvar, t PolyType) Kind {
+	return k
+}
+func (l KClass) unifyKind(kinds map[Tvar]Kind, r Kind) (Kind, Substitution, error) {
+	//TODO
+	return nil, nil, nil
+}
+func (k KClass) resolveType(map[Tvar]Kind) (Type, error) {
+	return nil, errors.New("KClass has no type")
+}
+func (k KClass) MonoType() (Type, bool) {
+	return nil, false
+}
+func (k KClass) resolvePolyType(map[Tvar]Kind) (PolyType, error) {
+	return nil, errors.New("KClass has no poly type")
+}
+
+type KRecord struct {
+	properties map[string]PolyType
+	lower      LabelSet
+	upper      LabelSet
+}
+
+func (k KRecord) String() string {
+	return fmt.Sprintf("{%v %v %v}", k.properties, k.lower, k.upper)
+}
+
+func (k KRecord) Invalid() bool {
+	for _, l := range k.lower {
+		t := k.properties[l]
+		_, ok := t.(invalid)
+		if ok {
+			return true
+		}
+	}
+	return false
+}
+
+func (k KRecord) substituteKind(tv Tvar, t PolyType) Kind {
+	properties := make(map[string]PolyType)
+	for k, f := range k.properties {
+		properties[k] = f.substituteType(tv, t)
+	}
+	return KRecord{
+		properties: properties,
+		upper:      k.upper.copy(),
+		lower:      k.lower.copy(),
+	}
+}
+func (k KRecord) freeVars(c *Constraints) TvarSet {
+	var fvs TvarSet
+	for _, f := range k.properties {
+		fvs = fvs.union(f.freeVars(c))
+	}
+	return fvs
+}
+
+func (l KRecord) unifyKind(kinds map[Tvar]Kind, k Kind) (kind Kind, _ Substitution, _ error) {
+	r, ok := k.(KRecord)
+	if !ok {
+		return nil, nil, fmt.Errorf("cannot unify record with %T", k)
+	}
+
+	// Merge properties building up a substitution
+	subst := make(Substitution)
+	properties := make(map[string]PolyType, len(l.properties)+len(r.properties))
+	for f, typL := range l.properties {
+		properties[f] = typL
+		typR, ok := r.properties[f]
+		if !ok {
+			continue
+		}
+		s, err := unifyTypes(kinds, typL, typR)
+		if err != nil {
+			properties[f] = invalid{}
+		}
+		subst.Merge(s)
+		properties[f] = subst.ApplyType(typL)
```

Comment:
Does this actually handle `invalid` correctly? Are these two statements supposed to be in an else, or does the returned substitution actually map `typL` to `invalid` if there's an error?

Reply:
Ah good catch, these are supposed to be in an else. I'll add a test case for the type error condition.